### PR TITLE
Fixes #4037: Paginate and Filter Orders Subscription on Admin Dashboard

### LINKED ISSUE #4037: Paginate and filter Orders Subscription on Admin Dashboard
## Issue Description
The `CustomPaginatedOrders` subscription doesn't paginate the query and pulls all the orders from the DB. When you have an Orders collection of over 700 records this resulted in an interrupted query and on the Orders Dashboard panel never loading the orders.

It throws the following error:
```
Exception in added observe/observeChanges callback: MongoError: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting. Aborting operation. Pass allowDiskUse:true to opt in. 
```

## Steps to Reproduce
Logged in as Admin.
1. Add over 700 orders to the database
2. Try to open the Orders menu from the Admin panel
3. See the eternal loading icon and the stack trace on the server

eternal-loading

## Possible Solution
Paginate and query on the `CustomPaginatedOrders` subscription method. In this case add a default pagination of a 100 records that matches the one used on the search filter.

### PR DESCRIPTION
**Pull request by glmaljkovich**
Resolves #4037 
Impact: **critical**  
Type: **bugfix|performance**

## Issue
The `CustomPaginatedOrders` subscription doesn't paginate the query and pulls all the orders from the DB. When you have an Orders collection of over 700 records this resulted in an interrupted query and on the Orders Dashboard panel never loading the orders.

It throws the following error:
```
Exception in added observe/observeChanges callback: MongoError: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting. Aborting operation. Pass allowDiskUse:true to opt in. 
```

## Solution
Paginate and query on the `PaginatedOrders` subscription method. 

## Steps to Reproduce
Context: Logged in as Admin
1. Add over 700 orders to the database
2. Try to open the Orders menu from the Admin panel
3. See the eternal loading icon and the stack trace on the server

## Breaking changes
Shouldn't have any. Added two optional parameters to one function

## Testing
1. Add over 700 records to the Orders collection.
2. Go to the Admin dashboard and open the Orders menu
3. You should see all the orders and filter them without any problem.

### CODE REVIEW

**Comment by zenweasel** [comment]
Also a couple of lint errors. I would just fix those but I can't push to your remote repo.

**Comment by glmaljkovich** [comment]
True, I overlooked the limit. I'll see what can be implemented to link the table pagination with the order subscription pagination.

Linter errors fixed

**Comment by zenweasel** [comment]
@glmaljkovich If you look through the docs for `react-table` I believe there's an example of how to do pagination

**Comment by glmaljkovich** [comment]
Yes, I'm working on that. I'll use the onFetchData or the table hooks to
pass the pagination parameters as props to the SubscriptionContainer and
refresh the Subscription as well as trigger the Orders.find()

I hope to have updates by the end of day.

**Comment by glmaljkovich** [comment]
Ok, now it should work mostly fine. I added:

- A publication to expose the orders collection size to the admin (using the Counts package)
- Dynamic page count based on collection size or search results
- Server side pagination
- Performance improvements on Order summary (Only watch one order)

It still needs to set the table on a loading state while its fetching data from the server.

**Comment by zenweasel** [comment]
Thanks so much! Will check it out on Monday

**Comment by zenweasel** [comment]
@glmaljkovich this works pretty well. I loaded 10k orders and was able to paginate back and forth. It was a little slow but no endless spinner so a **huge** improvement.

When I tried to go to a particular page (by entering the page number) however I got this error on the server:

```
Exception from sub CustomPaginatedOrders id Rt57tWdGNsHxc4K5c MongoError: Sort exceeded memory limit of 104857600 bytes, but did not opt in to external sorting. Aborting operation. Pass allowDiskUse:true to opt in.
    at Function.MongoError.create (/Users/brenthoover/.meteor/packages/npm-mongo/.2.2.34.10cdr9z.bl15l++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/error.js:31:11)
    at queryCallback (/Users/brenthoover/.meteor/packages/npm-mongo/.2.2.34.10cdr9z.bl15l++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/cursor.js:212:36)
    at /Users/brenthoover/.meteor/packages/npm-mongo/.2.2.34.10cdr9z.bl15l++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/connection/pool.js:469:18
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickDomainCallback (internal/process/next_tick.js:218:9)
```

Not sure if that's something with this code per-se or if I need to be configuring on my Mongo server if I have that many records. I will dig deeper and let you know what you find.

File: imports/plugins/core/orders/client/containers/orderSubscriptionContainer.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ import { Meteor } from "meteor/meteor";
 import { Orders, OrderSearch } from "/lib/collections";
 import { Components, composeWithTracker } from "@reactioncommerce/reaction-components";
 import OrderDashboard from "../components/orderDashboard";
+import { Counts } from "meteor/tmeasday:publish-counts";
```
Comment (zenweasel) [inline imports/plugins/core/orders/client/containers/orderSubscriptionContainer.js]: Import order

File: server/publications/collections/orders.js [hunk 2]
```diff
@@ -4,20 +4,21 @@ import { Roles } from "meteor/alanning:roles";
 import { ReactiveAggregate } from "./reactiveAggregate";
 import { MediaRecords, Orders } from "/lib/collections";
 import { Reaction } from "/server/api";
-
+import { Counts } from "meteor/tmeasday:publish-counts";
```
Comment (zenweasel) [inline server/publications/collections/orders.js]: Import order

**Comment by zenweasel** [comment]
Ok, it appears that error is (as it says) because if you are sorting large datasets (like we are with a large amount of orders) in a pipeline we need to specify `allowDiskUser: true`. [docs](https://docs.mongodb.com/manual/reference/command/aggregate/)

**Comment by glmaljkovich** [comment]
Yes, I had seen that flag. I was hoping we could make the query fast and
small enough to not have to rely on it.

But no problem. I'll add it to the aggregation on createAggregate. Another
thing I noticed is that I'm explicitly setting the manual flag on
SortedTable. But if other views use that table then it could cause
problems. Those views should be refactored anyway because they would suffer
from the same performance problem the Orders have.

**Comment by zenweasel** [comment]
@glmaljkovich Using that flag doesn't have to do with the efficiency of the query but the size of the dataset, so if you have large datasets I don't think it's avoidable. My concern is maybe there are some hosting environments where maybe that isn't allowed and maybe we should have some way to turn that option off in the settings

**Comment by glmaljkovich** [comment]
I propose adding the `MONGO_ALLOW_DISK_USE` env variable and leaving the option to override it on each call to ReactiveAggregate if you pass `allowDiskUse` inside the `options` parameter.

I fixed the manual flag

**Comment by zenweasel** [comment]
@glmaljkovich re: `allowDiskUse`. Yes, that's exactly what I was thinking. It's probably not a common thing but that way we aren't creating a potential issue for someone in a more limited hosting environment

**Comment by glmaljkovich** [comment]
Is there anything missing beside the change on imports order and the linter errors? I already added the env variable.

**Comment by zenweasel** [comment]
@glmaljkovich No, I will take a look at it again today. But we can't merge until linting errors are fixed however

File: imports/plugins/core/orders/client/components/orderDashboard.js [hunk 3]
```diff
@@ -17,7 +17,11 @@ class OrderDashboard extends Component {
     handleSelect: PropTypes.func,
     isLoading: PropTypes.object,
     multipleSelect: PropTypes.bool,
+    onPageChange: PropTypes.func,
+    onPageSizeChange: PropTypes.func,
     orders: PropTypes.array,
+    page: PropTypes.number,
```
Comment (zenweasel) [inline imports/plugins/core/orders/client/components/orderDashboard.js]: Could we rename this to `currentPage` to make it more clear what it is?

File: imports/plugins/core/orders/client/containers/orderSubscriptionContainer.js [hunk 4]
```diff
@@ -28,13 +29,21 @@ function composer(props, onData) {
       // being here means no search text is inputed or search was cleared, so reset any previous match
       delete query._id;
     }
-
-    const ordersSubscription = Meteor.subscribe("CustomPaginatedOrders", query);
+    const options = { limit: props.pageSize, skip: props.skip };
+    const ordersSubscription = Meteor.subscribe("CustomPaginatedOrders", query, options);
+    let total = Counts.get("orders-count");
+    if (props.searchQuery !== "") {
+      total = orderSearchResults.length;
```
Comment (zenweasel) [inline imports/plugins/core/orders/client/containers/orderSubscriptionContainer.js]: Could we rename this to something more clear that it's the total order count?

**Comment by glmaljkovich** [comment]
I went for replacing `PaginatedOrders` because I wasn't sure if the difference when getting the shopId matters. In any case it can be refactored later. Give it one last run in case I screwed up cjanging variable names. But it should be ready.

**Comment by glmaljkovich** [comment]
@zenweasel what are the next steps?

**Comment by zenweasel** [comment]
@glmaljkovich Sorry, just been swamped. Will retest one last time today and approve.

**Pull request merged**